server/task_executor_pitr_restore.go

```diff
@@ -0,0 +1,162 @@
+package server
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/bytebase/bytebase/api"
+	"github.com/bytebase/bytebase/common"
+	"github.com/bytebase/bytebase/plugin/db"
+	pluginmysql "github.com/bytebase/bytebase/plugin/db/mysql"
+	"github.com/bytebase/bytebase/plugin/db/util"
+	restoremysql "github.com/bytebase/bytebase/plugin/restore/mysql"
+	"github.com/bytebase/bytebase/store"
+	"go.uber.org/zap"
+)
+
+// NewPITRRestoreTaskExecutor creates a PITR restore task executor.
+func NewPITRRestoreTaskExecutor(logger *zap.Logger) TaskExecutor {
+	return &PITRRestoreTaskExecutor{
+		l: logger,
+	}
+}
+
+// PITRRestoreTaskExecutor is the PITR restore task executor.
+type PITRRestoreTaskExecutor struct {
+	l *zap.Logger
+}
+
+// RunOnce will run the PITR restore task executor once.
+func (exec *PITRRestoreTaskExecutor) RunOnce(ctx context.Context, server *Server, task *api.Task) (terminated bool, result *api.TaskRunResultPayload, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			panicErr, ok := r.(error)
+			if !ok {
+				panicErr = fmt.Errorf("%v", r)
+			}
+			exec.l.Error("PITRRestoreTaskExecutor PANIC RECOVER", zap.Error(panicErr), zap.Stack("stack"))
+			terminated = true
+			err = fmt.Errorf("encounter internal error when executing sql")
+		}
+	}()
+	exec.l.Info("Run PITR restore task", zap.String("task", task.Name))
+
+	payload := api.TaskDatabasePITRRestorePayload{}
+	if err := json.Unmarshal([]byte(task.Payload), &payload); err != nil {
+		return true, nil, fmt.Errorf("invalid PITR restore payload[%s], error[%w]", task.Payload, err)
+	}
+
+	return exec.pitrRestore(ctx, task, server)
+}
+
+func (exec *PITRRestoreTaskExecutor) pitrRestore(ctx context.Context, task *api.Task, server *Server) (terminated bool, result *api.TaskRunResultPayload, err error) {
+	schemaVersion := common.DefaultMigrationVersion()
+	migrationInfo, err := preMigration(ctx, exec.l, server, task, db.Baseline, "", schemaVersion, nil)
+	if err != nil {
+		exec.l.Error("failed in preMigration stage", zap.Error(err))
+		return true, nil, fmt.Errorf("failed in preMigration stage, error[%w]", err)
+	}
+
+	driver, err := getAdminDatabaseDriver(ctx, task.Instance, "", exec.l)
+	if err != nil {
+		return true, nil, err
+	}
+	defer driver.Close(ctx)
+
+	needsSetup, err := driver.NeedsSetupMigration(ctx)
```

Comment:
nit: we can do a refactoring later and move this check to a common place.

Reply:
Added TODO.